From 75c96d7319f39295eda38faee12e1205681cde8c Mon Sep 17 00:00:00 2001
From: Sathyam Panda <sathyam.panda@arm.com>
Date: Sat, 17 Dec 2022 20:19:51 +0000
Subject: [PATCH] drivers/arm: mhuv3: add MHUv3 driver

Add support of MHUv3 sender/receiver block initialization.
At initial stage provide only Doorbell Extension support.

Upstream-Status: Pending
Signed-off-by: Sathyam Panda <sathyam.panda@arm.com>
Signed-off-by: Peter Hoyes <Peter.Hoyes@arm.com>
---
 drivers/arm/mhu/mhu_v3_x.c  | 392 +++++++++++++++++++++++++++++++++
 include/drivers/arm/mhuv3.h | 426 ++++++++++++++++++++++++++++++++++++
 2 files changed, 818 insertions(+)
 create mode 100644 drivers/arm/mhu/mhu_v3_x.c
 create mode 100644 include/drivers/arm/mhuv3.h

diff --git a/drivers/arm/mhu/mhu_v3_x.c b/drivers/arm/mhu/mhu_v3_x.c
new file mode 100644
index 000000000..6ec3a8a8d
--- /dev/null
+++ b/drivers/arm/mhu/mhu_v3_x.c
@@ -0,0 +1,392 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <drivers/arm/mhuv3.h>
+#include <drivers/delay_timer.h>
+
+
+static void discover_mhu_v3_features(mhu_dev_t *dev,
+		mhu_v3_feat_t *mhu_v3_features, size_t feature_list_len,
+		mhu_frame_t frame_t)
+{
+	uint32_t feat_spt0, feat_spt1, feat;
+	uint32_t i, val;
+	struct mhu3_pbx_ctrl_page *pbx_ctrl_page;
+	struct mhu3_mbx_ctrl_page *mbx_ctrl_page;
+
+	dev->mhu_v3_features = mhu_v3_features;
+	if (frame_t == MHU_SENDER_FRAME) {
+		if (false == dev->is_initialized[MHU_SENDER_FRAME])
+			return;
+		pbx_ctrl_page = (struct mhu3_pbx_ctrl_page *)
+			(dev->pbx + MHU3_PBX_PBX_CTRL_PAGE_OFFSET);
+		feat_spt0 = pbx_ctrl_page->PBX_FEAT_SPT0;
+		feat_spt1 = pbx_ctrl_page->PBX_FEAT_SPT1;
+	} else if (frame_t == MHU_RECEIVER_FRAME) {
+		if (false == dev->is_initialized[MHU_RECEIVER_FRAME])
+			return;
+		mbx_ctrl_page = (struct mhu3_mbx_ctrl_page *)
+			(dev->mbx + MHU3_MBX_MBX_CTRL_PAGE_OFFSET);
+		feat_spt0 = mbx_ctrl_page->MBX_FEAT_SPT0;
+		feat_spt1 = mbx_ctrl_page->MBX_FEAT_SPT1;
+	} else {
+		WARN("Undefined MHUv3 Frame Type :%u\n", frame_t);
+		return;
+	}
+
+
+	INFO("MHUv3-%s: Supported Features\n", frame_t == MHU_SENDER_FRAME ?
+			"Postbox" : "Mailbox");
+
+	for (i = 0; i < feature_list_len; i++) {
+		feat = (mhu_v3_features[i].feat_reg ? feat_spt1 :
+				feat_spt0);
+		val = (feat & GENMASK(mhu_v3_features[i].feat_msb,
+					mhu_v3_features[i].feat_lsb)) >>
+			mhu_v3_features[i].feat_lsb;
+		if (val != mhu_v3_features[i].feat_inv_val[frame_t]) {
+			mhu_v3_features[i].feat_support[frame_t] = val;
+			mhu_v3_features[i].feat_valid[frame_t] = 1;
+			INFO("\t%s:%u\n", mhu_v3_features[i].feat_name, val);
+		}
+	}
+}
+
+static void mhu_v3_init_sender_frame_doorbell_ch(mhu_dev_t *dev)
+{
+	struct mhu3_pdbcw_page *pdbcw_page;
+	struct mhu3_pbx_ctrl_page *pbx_ctrl_page;
+	uint32_t ch;
+
+	pbx_ctrl_page = (struct mhu3_pbx_ctrl_page *)
+		(dev->pbx + MHU3_PBX_PBX_CTRL_PAGE_OFFSET);
+	pdbcw_page = (struct mhu3_pdbcw_page *)
+		(dev->pbx + MHU3_PBX_PDBCW_PAGE_OFFSET);
+
+	/* Read total number of supported channels */
+	dev->mhu_v3_pbx_db.num_dbch = READ_REG(pbx_ctrl_page->PBX_DBCH_CFG0,
+			DBCH_CFG0_NUM_DBCH_SHIFT, DBCH_CFG0_NUM_DBCH_MASK) + 1;
+	dev->mhu_v3_pbx_db.num_free_dbch = dev->mhu_v3_pbx_db.num_dbch;
+
+	for (ch = 0 ; ch < dev->mhu_v3_pbx_db.num_dbch; ch++) {
+		/* Clear the Transfer Acknowledge */
+		WRITE_REG(pdbcw_page[ch].PDBCW_INT_CLR,
+				PDBCW_INT_CLR_TFR_ACK,
+				PDBCW_INT_CLR_TFR_ACK_SHIFT,
+				PDBCW_INT_CLR_TFR_ACK_MASK);
+		/* Disable Transfer Acknowledge Interrupt */
+		WRITE_REG(pdbcw_page[ch].PDBCW_INT_EN,
+				~PDBCW_INT_EN_TFR_ACK,
+				PDBCW_INT_EN_TFR_ACK_SHIFT,
+				PDBCW_INT_EN_TFR_ACK_MASK);
+		/* Prevent channel contribution to PBX combined interrupt */
+		WRITE_REG(pdbcw_page[ch].PDBCW_CTRL,
+				~PDBCW_CTRL_PBX_COMB_EN,
+				PDBCW_CTRL_PBX_COMB_EN_SHIFT,
+				PDBCW_CTRL_PBX_COMB_EN_MASK);
+		dev->mhu_v3_pbx_db.dbch_state[ch] = FREE;
+	}
+
+}
+
+static void mhu_v3_init_receiver_frame_doorbell_ch(mhu_dev_t *dev)
+{
+	struct mhu3_mdbcw_page *mdbcw_page;
+	struct mhu3_mbx_ctrl_page *mbx_ctrl_page;
+	uint32_t ch;
+
+	mbx_ctrl_page = (struct mhu3_mbx_ctrl_page *)
+		(dev->mbx + MHU3_MBX_MBX_CTRL_PAGE_OFFSET);
+	mdbcw_page = (struct mhu3_mdbcw_page *)
+		(dev->mbx + MHU3_MBX_MDBCW_PAGE_OFFSET);
+
+	/* Read total number of supported channels */
+	dev->mhu_v3_mbx_num_dbch = READ_REG(mbx_ctrl_page->MBX_DBCH_CFG0,
+			DBCH_CFG0_NUM_DBCH_SHIFT, DBCH_CFG0_NUM_DBCH_MASK) + 1;
+
+	for (ch = 0; ch < dev->mhu_v3_mbx_num_dbch; ch++) {
+		/* Mask interrupts for all the flags to stop listening */
+		mdbcw_page[ch].MDBCW_MSK_SET = (uint32_t)MHU3_ALL_FLAGS_MASK;
+		/* Prevent channel contribution to MBX combined interrupt */
+		WRITE_REG(mdbcw_page[ch].MDBCW_CTRL,
+				~MDBCW_CTRL_MBX_COMB_EN,
+				MDBCW_CTRL_MBX_COMB_EN_SHIFT,
+				MDBCW_CTRL_MBX_COMB_EN_MASK);
+	}
+}
+
+int mhu_v3_sender_frame_init(mhu_dev_t *dev, uintptr_t base,
+		mhu_v3_feat_t *mhu_v3_features, size_t feature_list_len)
+{
+	struct mhu3_pbx_ctrl_page *pbx_ctrl_page;
+
+	if ((void *)base == NULL)
+		return -EFAULT;
+
+	dev->pbx = base;
+	pbx_ctrl_page = (struct mhu3_pbx_ctrl_page *)(dev->pbx +
+			MHU3_PBX_PBX_CTRL_PAGE_OFFSET);
+
+	if (READ_REG(pbx_ctrl_page->PBX_BLK_ID, PBX_BLK_ID_SHIFT,
+				PBX_BLK_ID_MASK) != MHU_SENDER_FRAME) {
+		ERROR("Not a Sender Frame, please check IO base address\n");
+		return -EINVAL;
+	}
+
+	WRITE_REG(pbx_ctrl_page->PBX_CTRL, PBX_CTRL_OP_REQ,
+			PBX_CTRL_OP_REQ_SHIFT, PBX_CTRL_OP_REQ_MASK);
+	WRITE_REG(pbx_ctrl_page->PBX_CTRL, ~PBX_CTRL_CH_OP_MSK_IGNORE,
+			PBX_CTRL_CH_OP_MSK_SHIFT, PBX_CTRL_CH_OP_MSK_MASK);
+
+	dev->product_id[MHU_SENDER_FRAME] = READ_REG(pbx_ctrl_page->PBX_IIDR,
+			IIDR_PRODUCT_ID_SHIFT, IIDR_PRODUCT_ID_MASK);
+	dev->variant[MHU_SENDER_FRAME] = READ_REG(pbx_ctrl_page->PBX_IIDR,
+			IIDR_VARIANT_SHIFT, IIDR_VARIANT_MASK);
+	dev->version[MHU_SENDER_FRAME].major = READ_REG(pbx_ctrl_page->PBX_AIDR,
+			AIDR_ARCH_MAJOR_REV_SHIFT, AIDR_ARCH_MAJOR_REV_MASK);
+	dev->version[MHU_SENDER_FRAME].minor = READ_REG(pbx_ctrl_page->PBX_AIDR,
+			AIDR_ARCH_MINOR_REV_SHIFT, AIDR_ARCH_MINOR_REV_MASK);
+
+	dev->is_initialized[MHU_SENDER_FRAME] = true;
+	INFO("MHUv3-Postbox: Product:%x Variant:%x Version:[%u:%u]\n",
+			dev->product_id[MHU_SENDER_FRAME],
+			dev->variant[MHU_SENDER_FRAME],
+			dev->version[MHU_SENDER_FRAME].major,
+			dev->version[MHU_SENDER_FRAME].minor);
+
+	/* Obtain MHUv3 Capabilities */
+	discover_mhu_v3_features(dev, mhu_v3_features, feature_list_len,
+			MHU_SENDER_FRAME);
+
+	/* Initialize Doorbell Channels */
+	if (mhu_v3_features[DBE_SPT].feat_valid[MHU_SENDER_FRAME])
+		mhu_v3_init_sender_frame_doorbell_ch(dev);
+
+	dev->global_owner_id_counter = 0;
+
+	return 0;
+}
+
+int mhu_v3_receiver_frame_init(mhu_dev_t *dev, uint64_t base,
+		mhu_v3_feat_t *mhu_v3_features, size_t feature_list_len)
+{
+	struct mhu3_mbx_ctrl_page *mbx_ctrl_page;
+
+	if ((void *)base == NULL)
+		return -EFAULT;
+
+	dev->mbx = base;
+	mbx_ctrl_page = (struct mhu3_mbx_ctrl_page *)(dev->mbx +
+			MHU3_MBX_MBX_CTRL_PAGE_OFFSET);
+
+	if (READ_REG(mbx_ctrl_page->MBX_BLK_ID, MBX_BLK_ID_SHIFT,
+				MBX_BLK_ID_MASK) != MHU_RECEIVER_FRAME) {
+		ERROR("Not a Receiver Frame, please check IO base address\n");
+		return -EINVAL;
+	}
+
+	WRITE_REG(mbx_ctrl_page->MBX_CTRL, MBX_CTRL_OP_REQ,
+			MBX_CTRL_OP_REQ_SHIFT, MBX_CTRL_OP_REQ_MASK);
+	WRITE_REG(mbx_ctrl_page->MBX_CTRL, ~MBX_CTRL_CH_OP_MSK_IGNORE,
+			MBX_CTRL_CH_OP_MSK_SHIFT, MBX_CTRL_CH_OP_MSK_MASK);
+
+	dev->product_id[MHU_RECEIVER_FRAME] = READ_REG(mbx_ctrl_page->MBX_IIDR,
+			IIDR_PRODUCT_ID_SHIFT, IIDR_PRODUCT_ID_MASK);
+	dev->variant[MHU_RECEIVER_FRAME] = READ_REG(mbx_ctrl_page->MBX_IIDR,
+			IIDR_VARIANT_SHIFT, IIDR_VARIANT_MASK);
+	dev->version[MHU_RECEIVER_FRAME].major = READ_REG(mbx_ctrl_page->MBX_AIDR,
+			AIDR_ARCH_MAJOR_REV_SHIFT, AIDR_ARCH_MAJOR_REV_MASK);
+	dev->version[MHU_RECEIVER_FRAME].minor = READ_REG(mbx_ctrl_page->MBX_AIDR,
+			AIDR_ARCH_MINOR_REV_SHIFT, AIDR_ARCH_MINOR_REV_MASK);
+
+	dev->is_initialized[MHU_RECEIVER_FRAME] = true;
+	INFO("MHUv3-Mailbox: Product:%x Variant:%x Version:[%u:%u]\n",
+			dev->product_id[MHU_RECEIVER_FRAME],
+			dev->variant[MHU_RECEIVER_FRAME],
+			dev->version[MHU_RECEIVER_FRAME].major,
+			dev->version[MHU_RECEIVER_FRAME].minor);
+
+	/* Obtain MHUv3 Capabilities */
+	discover_mhu_v3_features(dev, mhu_v3_features, feature_list_len,
+			MHU_RECEIVER_FRAME);
+
+	/* Initialize Doorbell Channels */
+	if (mhu_v3_features[DBE_SPT].feat_valid[MHU_RECEIVER_FRAME])
+		mhu_v3_init_receiver_frame_doorbell_ch(dev);
+
+	return 0;
+}
+
+/*
+ * This function will require num of channels as input and in response will
+ * derive a unique Owner ID and allocate channels to that Owner ID and return
+ * the same to caller which has to be provided by the caller while requesting
+ * for any MHU operation.
+ */
+int mhu_v3_sender_frame_reserve_doorbell_channels(mhu_dev_t *dev,
+		uint64_t *owner_id, uint32_t nums, uint32_t *channel_ids)
+{
+	mhu_v3_db_t *pbx = &dev->mhu_v3_pbx_db;
+	uint32_t ch, j;
+
+	if (!nums) {
+		ERROR("Invalid num of sender doorbell channels requested\n");
+		return -EINVAL;
+	}
+	if (nums > (pbx->num_free_dbch)) {
+		ERROR("Not enough free sender doorbell channels\n");
+		return -ENOSPC;
+	}
+
+	/*
+	 * TODO:
+	 * Use a standard method to generate unique ID everytime this function is
+	 * invoked. On multi core system at runtime bakery lock has to be used to
+	 * manage channel ownership in this function during while request to use
+	 * MHU arises from EL2 or EL1 software.
+	 */
+	dev->global_owner_id_counter++;
+	*owner_id = dev->global_owner_id_counter;
+
+	for (ch = 0, j = 0; (j < nums) && (ch < pbx->num_dbch); ch++) {
+
+		if (pbx->dbch_state[ch] == FREE) {
+			pbx->dbch_state[ch] = IN_USE;
+			pbx->num_free_dbch -= 1;
+			pbx->owner[ch] = *owner_id;
+			channel_ids[j] = ch;
+			j++;
+		}
+	}
+
+	return 0;
+}
+
+void mhu_v3_sender_frame_release_doorbell_channels(mhu_dev_t *dev,
+		uint32_t owner_id)
+{
+	mhu_v3_db_t *pbx = &dev->mhu_v3_pbx_db;
+	uint32_t ch;
+
+	for (ch = 0; ch < pbx->num_dbch; ch++) {
+
+		if ((pbx->owner[ch] == owner_id) &&
+				(pbx->dbch_state[ch] == IN_USE)) {
+			pbx->dbch_state[ch] = FREE;
+			pbx->num_free_dbch += 1;
+		}
+	}
+}
+
+int mhu_v3_sender_frame_write_doorbell_channels(mhu_dev_t *dev,
+		uint32_t owner_id, uint32_t nums_ch, mhu_v3_db_data_t *data)
+{
+	uint32_t i, retry = 30;
+	struct mhu3_pdbcw_page *pdbcw_page;
+	mhu_v3_db_t *pbx = &dev->mhu_v3_pbx_db;
+
+	pdbcw_page = (struct mhu3_pdbcw_page *)
+		(dev->pbx + MHU3_PBX_PDBCW_PAGE_OFFSET);
+
+	for (i = 0; i < nums_ch; i++) {
+		if (pbx->owner[data[i].ch] != owner_id) {
+			ERROR("Operation not permitted: Unknown Owner ID\n");
+			return -EPERM;
+		}
+		/* Wait for the status to reset */
+		while ((pdbcw_page[data[i].ch].PDBCW_ST & (data[i].flags)) && retry) {
+			mdelay(20);
+			retry--;
+		}
+		if (!retry) {
+			ERROR("MHUv3:%p Sender timeout waiting for DBCH:%u\n",
+					(void *)dev->pbx, data[i].ch);
+			return -EIO;
+		}
+
+		pdbcw_page[data[i].ch].PDBCW_SET = data[i].flags;
+	}
+
+	return 0;
+}
+
+int mhu_v3_sender_frame_read_ack_doorbell_channels(mhu_dev_t *dev,
+		uint32_t owner_id, uint32_t nums_ch, mhu_v3_db_data_t *data)
+{
+	uint32_t i;
+	bool status, curr_status = false;
+	struct mhu3_pdbcw_page *pdbcw_page;
+	mhu_v3_db_t *pbx = &dev->mhu_v3_pbx_db;
+
+	pdbcw_page = (struct mhu3_pdbcw_page *)
+		(dev->pbx + MHU3_PBX_PDBCW_PAGE_OFFSET);
+
+	for (i = 0; i < nums_ch; i++) {
+		if (pbx->owner[data[i].ch] != owner_id) {
+			ERROR("Operation not permitted: Unknown Owner ID\n");
+			return -EPERM;
+		}
+	}
+	/* Wait for all the channels */
+	while (!curr_status) {
+		status = true;
+		for (i = 0; i < nums_ch && (status == true); i++) {
+			/* Status is still set i.e. not acknowledged yet */
+			if (pdbcw_page[data[i].ch].PDBCW_ST & (data[i].flags))
+				status = status & false;
+		}
+		curr_status = status;
+		mdelay(20);
+	}
+
+	return 0;
+}
+
+int mhu_v3_receiver_read_doorbell_ch(mhu_dev_t *dev, uint32_t ch, uint32_t *val)
+{
+	uint32_t retry = 30;
+	struct mhu3_mdbcw_page *mdbcw_page;
+	int ret = 0;
+
+	mdbcw_page = (struct mhu3_mdbcw_page *)
+		(dev->mbx + MHU3_MBX_MDBCW_PAGE_OFFSET);
+	do {
+		*val = mdbcw_page[ch].MDBCW_ST;
+		if (*val)
+			break;
+		mdelay(20);
+		retry--;
+	} while (!retry);
+
+	if (*val == 0) {
+		ERROR("Unable to find data on DB ch:%u\n", ch);
+		ret = -EIO;
+	} else {
+		WRITE_REG(mdbcw_page[ch].MDBCW_CLR, *val, 0, MHU3_ALL_FLAGS_MASK);
+	}
+
+	return ret;
+}
+
+int find_device(uintptr_t base_addr, mhu_v3_device_t *devs, size_t len,
+		mhu_frame_t frame_type)
+{
+	uint32_t idx;
+	int dev_index = -1;
+
+	for (idx = 0; idx < len; idx++) {
+		if ((devs[idx].base == base_addr) &&
+			(devs[idx].frame_type == frame_type))
+			dev_index = idx;
+	}
+
+	return dev_index;
+}
diff --git a/include/drivers/arm/mhuv3.h b/include/drivers/arm/mhuv3.h
new file mode 100644
index 000000000..b744a9da6
--- /dev/null
+++ b/include/drivers/arm/mhuv3.h
@@ -0,0 +1,426 @@
+/*
+ * Copyright (c) 2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef MHUV3_H
+#define MHUV3_H
+
+#include <stdint.h>
+
+#include <drivers/arm/mhu.h>
+#include <plat/arm/common/arm_def.h>
+
+/* Read-only register */
+#define FWK_R const volatile
+/* Write-only register */
+#define FWK_W volatile
+/* Read/write register */
+#define FWK_RW volatile
+
+#define MHU3_MAX_DOORBELL_CHANNELS UINT32_C(128)
+#define SCP_MHU3_MBX_INT_EN (4U)
+#define PBX_DBCH_FLAG_POS 0
+
+#define MHU3_MASK_GENERATE(BITSTART, LENGTH) \
+	(((1 << (LENGTH)) - 1) << ((BITSTART) + 1 - (LENGTH)))
+
+#define MHU3_MASKED_RECOVER(MASKED, BITSTART, LENGTH) \
+	((MASKED) >> ((BITSTART) + 1 - (LENGTH)))
+
+#define CLEAR_REG(reg, mask)	\
+	{ reg = reg & (~mask); }
+
+#define WRITE_REG(reg, val, shift, mask)	\
+	{				\
+		CLEAR_REG(reg, mask)	\
+		reg |= ((val << shift) & mask);	\
+	}
+#define	READ_REG(reg, shift, mask)	((reg & mask) >> shift)
+
+/*
+ * Postbox CTRL --> PBX.PBX_CTRL_page.PBX_CTRL
+ */
+#define PBX_CTRL_OP_REQ_MASK		BIT(0)
+#define PBX_CTRL_CH_OP_MSK_MASK		BIT(1)
+#define PBX_CTRL_OP_REQ_SHIFT		0
+#define PBX_CTRL_CH_OP_MSK_SHIFT	1
+#define PBX_CTRL_CH_OP_MSK_IGNORE	0x1U
+#define PBX_CTRL_OP_REQ			0x1U
+
+/*
+ * Postbox Block Identifier --> PBX.PBX_CTRL_page.PBX_BLK_ID
+ */
+#define PBX_BLK_ID_MASK			GENMASK(3, 0)
+#define PBX_BLK_ID_SHIFT		0
+
+/*
+ * Mailbox Block Identifier --> MBX.MBX_CTRL_page.MBX_BLK_ID
+ */
+#define MBX_BLK_ID_MASK			GENMASK(3, 0)
+#define MBX_BLK_ID_SHIFT		0
+
+/*
+ * Mailbox CTRL --> MBX.MBX_CTRL_page.MBX_CTRL
+ */
+#define MBX_CTRL_OP_REQ_MASK		BIT(0)
+#define MBX_CTRL_CH_OP_MSK_MASK		BIT(1)
+#define MBX_CTRL_OP_REQ_SHIFT		0
+#define MBX_CTRL_CH_OP_MSK_SHIFT	1
+#define MBX_CTRL_CH_OP_MSK_IGNORE	0x1U
+#define MBX_CTRL_OP_REQ			0x1U
+
+/*
+ * Postbox IIDR --> PBX.PBX_CTRL_page.PBX_IIDR
+ * Mailbox IIDR --> MBX.MBX_CTRL_page.MBX_IIDR
+ */
+#define	IIDR_IMPLEMENTER_MASK		GENMASK(11, 0)
+#define	IIDR_REVISION_MASK		GENMASK(15, 12)
+#define	IIDR_VARIANT_MASK		GENMASK(19, 16)
+#define	IIDR_PRODUCT_ID_MASK		GENMASK(31, 20)
+#define	IIDR_IMPLEMENTER_SHIFT		0
+#define	IIDR_REVISION_SHIFT		12
+#define	IIDR_VARIANT_SHIFT		16
+#define	IIDR_PRODUCT_ID_SHIFT		20
+
+/*
+ * Postbox AIDR --> PBX.PBX_CTRL_page.PBX_AIDR
+ * Mailbox AIDR --> MBX.MBX_CTRL_page.MBX_AIDR
+ */
+#define AIDR_ARCH_MINOR_REV_MASK	GENMASK(3, 0)
+#define AIDR_ARCH_MAJOR_REV_MASK	GENMASK(7, 4)
+#define AIDR_ARCH_MINOR_REV_SHIFT	0
+#define AIDR_ARCH_MAJOR_REV_SHIFT	4
+
+/*
+ * Postbox DBCH_CFG0 --> PBX.PBX_CTRL_page.PBX_DBCH_CFG0
+ * Mailbox DBCH_CFG0 --> MBX.MBX_CTRL_page.MBX_DBCH_CFG0
+ */
+#define	DBCH_CFG0_NUM_DBCH_MASK		GENMASK(7, 0)
+#define	DBCH_CFG0_NUM_DBCH_SHIFT	0
+#define	MH3_MAX_NUM_DBCH		U(128)
+#define	MH3_MAX_NUM_FCH			U(1024)
+
+/*
+ * Postbox PDBCW<n>_CTRL --> PBX.PDBCW_page.PDBCW_CTRL
+ */
+#define PDBCW_CTRL_PBX_COMB_EN_MASK	BIT(0)
+#define PDBCW_CTRL_PBX_COMB_EN_SHIFT	0
+#define PDBCW_CTRL_PBX_COMB_EN		0x1U
+
+/*
+ * Postbox PDBCW<n>_INT_EN --> PBX.PDBCW_page.PDBCW_INT_EN
+ */
+#define	PDBCW_INT_EN_TFR_ACK_MASK	BIT(0)
+#define	PDBCW_INT_EN_TFR_ACK_SHIFT	0
+#define	PDBCW_INT_EN_TFR_ACK		0x1U
+
+/*
+ * Postbox PDBCW<n>_INT_CLR --> PBX.PDBCW_page.PDBCW_INT_CLR
+ */
+#define	PDBCW_INT_CLR_TFR_ACK_MASK	BIT(0)
+#define	PDBCW_INT_CLR_TFR_ACK_SHIFT	0
+#define	PDBCW_INT_CLR_TFR_ACK		0x1U
+
+/*
+ * Mailbox MDBCW<n>_CTRL --> MBX.MDBCW_page.MDBCW_CTRL
+ */
+#define MDBCW_CTRL_MBX_COMB_EN_MASK	BIT(0)
+#define MDBCW_CTRL_MBX_COMB_EN_SHIFT	0
+#define MDBCW_CTRL_MBX_COMB_EN		0x1U
+
+#define MHU3_NUM_DB_FLAGS		32U
+#define MHU3_ALL_FLAGS_MASK		GENMASK(MHU3_NUM_DB_FLAGS - 1, 0)
+
+/*
+ * Offset
+ */
+#define	MHU3_PBX_PBX_CTRL_PAGE_OFFSET		UINT32_C(0x0)
+#define	MHU3_PBX_PDBCW_PAGE_OFFSET		UINT32_C(0x00001000)
+#define	MHU3_PBX_PFFCW_PAGE_OFFSET		UINT32_C(0x00002000)
+#define	MHU3_PBX_PFCW_PAGE_OFFSET		UINT32_C(0x00003000)
+#define	MHU3_PBX_PBX_IMPL_DEF_PAGE_OFFSET	UINT32_C(0x0000F000)
+
+#define	MHU3_MBX_MBX_CTRL_PAGE_OFFSET		UINT32_C(0x0)
+#define	MHU3_MBX_MDBCW_PAGE_OFFSET		UINT32_C(0x00001000)
+#define	MHU3_MBX_MFFCW_PAGE_OFFSET		UINT32_C(0x00002000)
+#define	MHU3_MBX_MFCW_PAGE_OFFSET		UINT32_C(0x00003000)
+#define	MHU3_MBX_MBX_IMPL_DEF_PAGE_OFFSET	UINT32_C(0x0000F000)
+
+/*!
+ * \brief MHU3 Postbox(PBX) Register Definitions
+ */
+struct mhu3_pbx_ctrl_page {
+	/*! Postbox control page */
+	FWK_R uint32_t PBX_BLK_ID;
+	uint8_t RESERVED1[0x10 - 0x04];
+	FWK_R uint32_t PBX_FEAT_SPT0;
+	FWK_R uint32_t PBX_FEAT_SPT1;
+	uint8_t RESERVED2[0x20 - 0x18];
+	FWK_R uint32_t PBX_DBCH_CFG0;
+	uint8_t RESERVED3[0x30 - 0x24];
+	FWK_R uint32_t PBX_FFCH_CFG0;
+	uint8_t RESERVED4[0x40 - 0x34];
+	FWK_R uint32_t PBX_FCH_CFG0;
+	uint8_t RESERVED5[0x50 - 0x44];
+	FWK_R uint32_t PBX_DCH_CFG0;
+	uint8_t RESERVED6[0x100 - 0x54];
+	FWK_RW uint32_t PBX_CTRL;
+	uint8_t RESERVED7[0x150 - 0x104];
+	FWK_RW uint32_t PBX_DMA_CTRL;
+	FWK_R uint32_t PBX_DMA_ST;
+	FWK_RW uint64_t PBX_DMA_CDL_BASE;
+	FWK_RW uint32_t PBX_DMA_CDL_PROP;
+	uint8_t RESERVED8[0x400 - 0x164];
+	FWK_R uint32_t PBX_DBCH_INT_ST[(0x410 - 0x400) >> 2];
+	FWK_R uint32_t PBX_FFCH_INT_ST[(0x430 - 0x410) >> 2];
+	FWK_R uint32_t PBX_DCH_INT_ST;
+	uint8_t RESERVED9[0xFC8 - 0x434];
+	FWK_R uint32_t PBX_IIDR;
+	FWK_R uint32_t PBX_AIDR;
+	FWK_R uint32_t IMPL_DEF_ID[4 * 11];
+};
+
+struct mhu3_pdbcw_page {
+	FWK_R uint32_t PDBCW_ST;
+	uint8_t RESERVED1[0xC - 0x4];
+	FWK_W uint32_t PDBCW_SET;
+	FWK_R uint32_t PDBCW_INT_ST;
+	FWK_W uint32_t PDBCW_INT_CLR;
+	FWK_RW uint32_t PDBCW_INT_EN;
+	FWK_RW uint32_t PDBCW_CTRL;
+};
+struct mhu3_mbx_ctrl_page {
+	/*! Mailbox control page */
+	FWK_R uint32_t MBX_BLK_ID;
+	uint8_t RESERVED1[0x10 - 0x04];
+	FWK_R uint32_t MBX_FEAT_SPT0;
+	FWK_R uint32_t MBX_FEAT_SPT1;
+	uint8_t RESERVED2[0x20 - 0x18];
+	FWK_R uint32_t MBX_DBCH_CFG0;
+	uint8_t RESERVED3[0x30 - 0x24];
+	FWK_R uint32_t MBX_FFCH_CFG0;
+	uint8_t RESERVED4[0x40 - 0x34];
+	FWK_R uint32_t MBX_FCH_CFG0;
+	uint8_t RESERVED5[0x50 - 0x44];
+	FWK_R uint32_t MBX_DCH_CFG0;
+	uint8_t RESERVED6[0x100 - 0x54];
+	FWK_RW uint32_t MBX_CTRL;
+	uint8_t RESERVED7[0x140 - 0x104];
+	FWK_RW uint32_t MBX_FCH_CTRL;
+	FWK_RW uint32_t MBX_FCG_INT_EN;
+	uint8_t RESERVED8[0x150 - 0x148];
+	FWK_RW uint32_t MBX_DMA_CTRL;
+	FWK_R uint32_t MBX_DMA_ST;
+	FWK_RW uint64_t MBX_DMA_CDL_BASE;
+	FWK_RW uint32_t MBX_DMA_CDL_PROP;
+	uint8_t RESERVED9[0x400 - 0x164];
+	FWK_R uint32_t MBX_DBCH_INT_ST[(0x410 - 0x400) >> 2];
+	FWK_R uint32_t MBX_FFCH_INT_ST[(0x420 - 0x410) >> 2];
+	FWK_R uint32_t MBX_FCG_INT_ST;
+	uint8_t RESERVED10[0x430 - 0x424];
+	FWK_R uint32_t MBX_FCH_GRP_INT_ST[(0x4B0 - 0x430) >> 2];
+	FWK_R uint32_t MBX_DCH_INT_ST;
+	uint8_t RESERVED11[0xFC8 - 0x4B4];
+	FWK_R uint32_t MBX_IIDR;
+	FWK_R uint32_t MBX_AIDR;
+	FWK_R uint32_t IMPL_DEF_ID[4 * 11];
+};
+
+struct mhu3_mdbcw_page {
+	FWK_R uint32_t MDBCW_ST;
+	FWK_R uint32_t MDBCW_ST_MSK;
+	FWK_RW uint32_t MDBCW_CLR;
+	uint8_t RESERVED1[0x10 - 0x0C];
+	FWK_R uint32_t MDBCW_MSK_ST;
+	FWK_RW uint32_t MDBCW_MSK_SET;
+	FWK_RW uint32_t MDBCW_MSK_CLR;
+	FWK_RW uint32_t MDBCW_CTRL;
+};
+
+typedef struct mhu_v3_db_data {
+	uint32_t ch;
+	uint8_t flags;
+} mhu_v3_db_data_t;
+
+typedef enum mhu_v3_cap_id {
+	DBE_SPT,
+	FE_SPT,
+	FCE_SPT,
+	TZE_SPT,
+	RME_SPT,
+	RASE_SPT,
+	AUTO_OP_SPT
+} mhu_v3_cap_id_t;
+
+
+typedef enum feat_spt {
+	FEAT_SPT0,
+	FEAT_SPT1
+} feat_spt_t;
+
+typedef enum mhu_frame {
+	MHU_SENDER_FRAME,
+	MHU_RECEIVER_FRAME,
+	MHU_MAX_FRAME_TYPES
+} mhu_frame_t;
+
+typedef enum mhu_v3_dbch_state {
+	INVALID,
+	FREE,
+	IN_USE
+} mhu_v3_dbch_state_t;
+/*
+ * MHU capability structure
+ */
+typedef struct mhu_v3_feat {
+	char *feat_name;
+	bool feat_valid[MHU_MAX_FRAME_TYPES];
+	int32_t feat_support[MHU_MAX_FRAME_TYPES];
+	int32_t feat_inv_val[MHU_MAX_FRAME_TYPES];
+	uint32_t feat_msb;
+	uint32_t feat_lsb;
+	feat_spt_t feat_reg;
+} mhu_v3_feat_t;
+
+typedef struct mhu_version {
+	uint8_t major;
+	uint8_t minor;
+} mhu_version_t;
+
+typedef struct mhu_v3_db {
+	uint32_t num_dbch;
+	uint32_t num_free_dbch;
+	mhu_v3_dbch_state_t dbch_state[MH3_MAX_NUM_DBCH];
+	uint64_t owner[MH3_MAX_NUM_DBCH];
+} mhu_v3_db_t;
+
+typedef struct mhu_dev {
+	uint64_t global_owner_id_counter;
+	mhu_version_t version[MHU_MAX_FRAME_TYPES];
+	uint32_t product_id[MHU_MAX_FRAME_TYPES];
+	uint32_t variant[MHU_MAX_FRAME_TYPES];
+	bool is_initialized[MHU_MAX_FRAME_TYPES];
+	uintptr_t pbx;
+	uintptr_t mbx;
+	mhu_v3_feat_t *mhu_v3_features;
+	mhu_v3_db_t mhu_v3_pbx_db;
+	uint32_t mhu_v3_mbx_num_dbch;
+} mhu_dev_t;
+
+typedef struct mhu_v3_device {
+	uintptr_t base;
+	uintptr_t mem_base;
+	mhu_dev_t *dev;
+	mhu_v3_feat_t *feat_list;
+	size_t feat_list_sz;
+	mhu_frame_t frame_type;
+} mhu_v3_device_t;
+
+/* Fast Channel Configuration 0 Number of Fast Channels Start Bit */
+#define MHU3_FCH_CFG0_NUM_FCH_BITSTART 9
+/* Fast Channel Configuration 0 Number of Fast Channels Length */
+#define MHU3_FCH_CFG0_NUM_FCH_LEN 10
+/* Fast Channel Configuration 0 Number of Fast Channels Mask */
+#define MHU3_FCH_CFG0_NUM_FCH_MASK \
+	(MHU3_MASK_GENERATE( \
+		MHU3_FCH_CFG0_NUM_FCH_BITSTART, MHU3_FCH_CFG0_NUM_FCH_LEN))
+
+/* Fast Channel Configuration 0 Number of Fast Channel Groups Start Bit */
+#define MHU3_FCH_CFG0_NUM_FCG_BITSTART 15
+/* Fast Channel Configuration 0 Number of Fast Channel Groups Length */
+#define MHU3_FCH_CFG0_NUM_FCG_LEN 5
+/* Fast Channel Configuration 0 Number of Fast Channel Groups Mask */
+#define MHU3_FCH_CFG0_NUM_FCG_MASK \
+	(MHU3_MASK_GENERATE( \
+		MHU3_FCH_CFG0_NUM_FCG_BITSTART, MHU3_FCH_CFG0_NUM_FCG_LEN))
+
+/*
+ * Fast Channel Configuration 0 Number of Fast Channels per Fast Channel Group
+ * Start Bit
+ */
+#define MHU3_FCH_CFG0_NUM_FCH_PER_GRP_BITSTART 20
+
+/*
+ * Fast Channel Configuration 0 Number of Fast Channels per Fast Channel Group
+ * Length
+ */
+#define MHU3_FCH_CFG0_NUM_FCH_PER_GRP_LEN 5
+
+/*
+ * Fast Channel Configuration 0 Number of Fast Channels per Fast Channel Group
+ * Mask
+ */
+#define MHU3_FCH_CFG0_NUM_FCH_PER_GRP_MASK \
+	(MHU3_MASK_GENERATE( \
+		MHU3_FCH_CFG0_NUM_FCH_PER_GRP_BITSTART, \
+		MHU3_FCH_CFG0_NUM_FCH_PER_GRP_LEN))
+
+/* Fast Channel Configuration 0 Fast Channel Word Size Start Bit */
+#define MHU3_FCH_CFG0_FCH_WS_BITSTART 28
+/* Fast Channel Configuration 0 Fast Channel Word Size Length */
+#define MHU3_FCH_CFG0_FCH_WS_LEN 8
+/* Fast Channel Configuration 0 Fast Channel Word Size Mask */
+#define MHU3_FCH_CFG0_FCH_WS_MASK \
+	(MHU3_MASK_GENERATE( \
+		MHU3_FCH_CFG0_FCH_WS_BITSTART, MHU3_FCH_CFG0_FCH_WS_LEN))
+
+
+/* Doorbell Flags */
+#define	MHU3_PBX_DCH_FLAG_SCMI		BIT_32(0)
+#define	MHU3_PBX_DCH_FLAG_RSS_COMMS	BIT_32(0)
+
+/* Base addresses of the NS AP<-->SCP MHUv3 */
+#define	AP2SCP_NS_PBX		0x2A900000
+#define	AP2SCP_NS_MBX		0x2A910000
+
+/* Base addresses of the S AP<-->SCP MHUv3 */
+#define	AP2SCP_S_PBX		0x2A920000
+#define	AP2SCP_S_MBX		0x2A930000
+
+/* Base addresses of the Root AP<-->SCP MHUv3 */
+#define	AP2SCP_ROOT_PBX		0x2A940000
+#define	AP2SCP_ROOT_MBX		0x2A950000
+
+/* Base addresses of the NS AP<-->RSS MHUv3 */
+#define	AP2RSS_NS_PBX		0x2AB00000
+#define	AP2RSS_NS_MBX		0x2AB10000
+
+/* Base addresses of the S AP<-->RSS MHUv3 */
+#define	AP2RSS_S_PBX		0x2AB20000
+#define	AP2RSS_S_MBX		0x2AB30000
+
+/* Base addresses of the Realm AP<-->RSS MHUv3 */
+#define	AP2RSS_REALM_PBX	0x2AB60000
+#define	AP2RSS_REALM_MBX	0x2AB70000
+
+/* Base addresses of the Root AP<-->RSS MHUv3 */
+#define	AP2RSS_ROOT_PBX		0x2AB40000
+#define	AP2RSS_ROOT_MBX		0x2AB50000
+
+
+extern int find_device(uintptr_t base_addr, mhu_v3_device_t *devs, size_t len,
+		mhu_frame_t frame_type);
+
+extern int mhu_v3_sender_frame_init(mhu_dev_t *dev, uintptr_t base,
+		mhu_v3_feat_t *mhu_v3_features, size_t feature_list_len);
+
+extern int mhu_v3_receiver_frame_init(mhu_dev_t *dev, uint64_t base,
+		mhu_v3_feat_t *mhu_v3_features, size_t feature_list_len);
+
+extern int mhu_v3_sender_frame_reserve_doorbell_channels(mhu_dev_t *dev,
+		uint64_t *owner_id, uint32_t nums, uint32_t *channel_ids);
+
+extern void mhu_v3_sender_frame_release_doorbell_channels(mhu_dev_t *dev,
+		uint32_t owner_id);
+
+extern int mhu_v3_sender_frame_write_doorbell_channels(mhu_dev_t *dev,
+		uint32_t owner_id, uint32_t nums_ch, mhu_v3_db_data_t *data);
+
+extern int mhu_v3_sender_frame_read_ack_doorbell_channels(mhu_dev_t *dev,
+		uint32_t owner_id, uint32_t nums_ch, mhu_v3_db_data_t *data);
+
+extern int mhu_v3_receiver_read_doorbell_ch(mhu_dev_t *dev, uint32_t ch,
+		uint32_t *val);
+#endif /* MHUV3 */
-- 
2.34.1

